crates/shared/src/event_handling.rs

```diff
@@ -286,19 +290,26 @@ where
 
     async fn update_events_from_latest_blocks(
         &mut self,
-        blocks: &[BlockNumberHash],
+        latest_blocks: &[BlockNumberHash],
         is_reorg: bool,
     ) -> Result<()> {
-        let (blocks, events) = self.past_events_by_block_hashes(blocks).await;
-        track_block_range(&format!("range_{}", blocks.len()));
-        tracing::debug!(
-            "final blocks for updating: {:?} - {:?}",
-            blocks.first(),
-            blocks.last()
+        anyhow::ensure!(
+            !latest_blocks.is_empty(),
+            "entered update events with empty block list"
         );
-        if blocks.is_empty() {
-            return Ok(());
+        let (blocks, events) = self.past_events_by_block_hashes(latest_blocks).await;
```

Comment:
Not 100% sure I follow the question.
We enter the `update_events_from_latest_blocks` with full list of blocks that we need update for in current iteration.
Then we potentially filter out some of those blocks for which we could not get events (for example, `latest_blocks` contains 3 blocks, but last one failed to get events because that block does not exist because of the node propagation issue).
Only then we do a `storage + last_handled_blocks` update, on a partial list of blocks or full list of blocks, in order to get to the most latest state we can.